internal/pkg/gateway/api.go

```diff
@@ -427,51 +428,111 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
-	for _, index := range rand.Perm(len(orderers)) {
-		orderer := orderers[index]
-		logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress)
-
-		var response *ab.BroadcastResponse
-		var err error
-		done := make(chan struct{})
-		go func() {
-			defer close(done)
-			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
-			defer cancel()
-
-			response, err = gs.broadcast(ctx, orderer, txn)
-		}()
-		select {
-		case <-done:
-			// Broadcast completed normally
-		case <-ctx.Done():
-			// Overall submit timeout expired
-			logger.Warn("Submit call timed out while broadcasting to ordering service")
-			return nil, newRpcError(codes.DeadlineExceeded, "submit timeout expired while broadcasting to ordering service")
+	config := gs.getChannelConfig(request.ChannelId)
+	if config.ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		// For BFT, we send transaction to ALL orderers
+		waitCh := make(chan *gp.ErrorDetail, len(orderers))
+		for _, o := range orderers {
+			go func(ord *orderer) {
+				logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+				ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
+				defer cancel()
+				response, err := gs.broadcast(ctx, ord, txn)
+
+				if err != nil {
+					logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "err", err)
+					waitCh <- errorDetail(ord.endpointConfig, err.Error())
+				} else if status := response.GetStatus(); status == common.Status_SUCCESS {
+					logger.Infow("Successful response from orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+					waitCh <- nil
+				} else {
+					logger.Warnw("Unsuccessful response sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "status", status, "info", response.GetInfo())
+					waitCh <- errorDetail(ord.endpointConfig, fmt.Sprintf("received unsuccessful response from orderer: %s", common.Status_name[int32(status)]))
+				}
+			}(o)
 		}
-
-		if err != nil {
-			errDetails = append(errDetails, errorDetail(orderer.endpointConfig, err.Error()))
-			logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress, "err", err)
-			continue
+		n := len(orderers)
+		ordererConfig, ok := config.OrdererConfig()
+		if ok {
+			n = len(ordererConfig.Consenters())
 		}
-
-		status := response.GetStatus()
-		if status == common.Status_SUCCESS {
-			return &gp.SubmitResponse{}, nil
+		f := (float64(n) - 1) / 3
+		quorum := int(math.Ceil(2*f + 1))
+		successes := 0
+		failures := 0
+	loop:
+		for i := 0; i < len(orderers); i++ {
+			select {
+			case osnErr := <-waitCh:
+				// Broadcast completed normally
+				if osnErr != nil {
+					errDetails = append(errDetails, osnErr)
+					failures++
+					if failures > n-quorum {
+						break loop
```

Comment:
I guess I don't quite understand.
If you come out here, what happens next?
What's the next scenario?

Reply:
If it finds that it's received too many failures to every attain 2f+1 successes, then bail out and return error to the client.  In this case the error is the gRPC code.Unavailable which hints to the client application that they could retry.  The client app might choose to re-submit the transaction with the same txID (which can be done via the finer-grained API).  If it turns out that the original transaction was processed successfully, then the one that was resent will be rejected as a duplicate.  I'm not sure the gateway can do any better than that.
Having said all that, I think the condition should be `failures > len(orderers) - quorum`.  I'll make that change.

Reply:
Can you tell me where I can see the "finer-grained API"?

Reply:
If you look at the code for `SubmitTransaction()`, you'll see that it's just syntactic sugar for a sequence of method calls at the level of `NewProposal()` ... `Endorse()` ... `Submit()` ... `Status()`.
So you could write your code to use these instead, and put in extra logic between each step depending on your requirements.  In this case, you might want to check the error returned by `Submit()`, and then resend to the orderer by calling `Submit()` again on the same transaction object (which will have the same txID).

https://github.com/hyperledger/fabric-gateway/blob/main/pkg/client/contract.go#L97
https://github.com/hyperledger/fabric-gateway/blob/main/pkg/client/contract.go#L169

Reply:
@andrew-coleman thanks

Reply:
I still think it will increase the number of false positives to return an error.

If a raft client sends a transaction to hlf+smartbft via sdk, there is a 95% chance it will be accepted and executed. This is equivalent to the fact that the client, after sending a request to all orders, will get only one positive response.

What does a positive or negative response from the orderer tell us? Only that the orderer has put the transaction into RequestPool, and not executed it. The execution is still a long way off.

In a distributed system, network failures are normal. And from the fact that there are failures between the client and the cluster nodes, it is premature to conclude that there are the same failures inside the cluster.